saw-remote-api/src/SAWServer/Data/SetupValue.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+module SAWServer.Data.SetupValue (CrucibleSetupVal) where
+
+import Control.Applicative
+import Data.Aeson (FromJSON(..), withObject, withText, (.:))
+
+import SAWServer
+
+data SetupValTag
+  = TagServerValue
+  | TagCryptol
+  | TagArrayValue
+  | TagFieldLValue
+  | TagElemLValue
+  | TagGlobalInit
+  | TagGlobalLValue
+
+instance FromJSON SetupValTag where
+  parseJSON =
+    withText "tag for setup value"$
+    \case
+      "saved" -> pure TagServerValue
+      "Cryptol" -> pure TagCryptol
+      "array value" -> pure TagArrayValue
+      "field lvalue" -> pure TagFieldLValue
+      "element lvalue" -> pure TagElemLValue
+      "global initializer" -> pure TagGlobalInit
+      "global lvalue" -> pure TagGlobalLValue
+      _ -> empty
+
+instance FromJSON cryptolExpr => FromJSON (CrucibleSetupVal cryptolExpr) where
+  parseJSON v =
+    nullValue v <|> fromObject v
+
+    where
+      nullValue = withText "setup value text" $
+        \case
+          "null" -> pure NullValue
```

Comment:
Just checking - do we want a protocol that uses the constant string `"null"` in a wire format with a constant named `null`? This seems like it might be a source of annoying bugs and misunderstandings. I don't think a JSON `null` makes sense here, but perhaps it should be some other string? And is `setup value text` what we want to see in parse errors?

Reply:
This is a remnant from old code that's just showing up as added because it's indented more. I was actually wondering about it, too. It's not only potentially confusing text, but also somewhat inconsistent with the other setup values. If you don't know of any reason to leave it how it is, I'll change it to make it more consistent and less confusing.

Reply:
The new version is much nicer, both as code and as interface.